flocker/node/agents/blockdevice.py

```diff
@@ -1489,6 +1649,17 @@ def is_existing_block_device(dataset_id, path):
             ).write(_logger)
             return False
 
+        # The state discovery method is basically the wrong place to do the
+        # filesystem upgrade.  Sadly, determining the necessity of each of
+        # these upgrades depends on more state than we can easily convey to
+        # `calculate_changes` (the technical term for this situation might be
+        # "design flaw").  Since we really do want to do these upgrades and
+        # we'd like to do them with a minimum of code churn, sticking this
+        # upgrade step into the discovery method makes the most sense.  This
+        # will be another datapoint to consider when revamping the way this
+        # IDeployer works.
+        self._upgrade_1_0_0_filesystem(api, volumes)
```

Comment:
:question: Seems like quite a lot of extra work, to be performing this check every time we loop for ever more, regardless of whether the volumes have already been upgraded.

 * Is it worth adding a `flocker-version` stamp or something to the volume metadata?
 * And maybe only performing this upgrade step when the agent starts up?
 * And what happens if the volume is mounted and can't be upgraded? Then the following code that calls `get_device_dataset_id` will fail every time....I think.